src/lowering/desugar.ml

```diff
@@ -239,6 +242,77 @@ and lexp' = function
   | S.IdxE (e1, e2) -> I.IdxLE (exp e1, exp e2)
   | _ -> raise (Invalid_argument ("Unexpected expression as lvalue"))
 
+and rewrite_for_to_while p arr proj c0 c1 c2 e1 e2 =
```

Comment:
I think this code would be much simpler and maintainable if you could use the helpers in Construct.ml that compute the types and effect for you. Helpers like Construct.letP,  letD, callE, blockE etc. Is there any reason you can't?

Reply:
I can, and it turns out to be much less hassle. I think when I started out this way I shouldn't have ripped the logic into two parts (in two source files), then I wouldn't have felt the urge to do the transformation in syntax-land only. PTAL.